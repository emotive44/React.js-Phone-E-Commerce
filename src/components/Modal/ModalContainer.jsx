import styled from 'styled-components';

const ModalContainer = styled.div`
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    display: flex;
    position: fixed;
    align-items: center;
    justify-content: center;
    background: rgba(0, 0, 0, 0.5);

    #modal {
        background: var(--mainWhite);
        border-radius: 1rem;
    }  
    .exit {
        position: absolute;
        top:0;
        right: 0;
        background: var(--mainYellow);
        border: none;
        outline: none;
        transition: all 0.6s linear;
        padding: 0 0.5rem;
        border-radius: 0 0.5rem 0 0 ;
        font-size: 2rem;
    }
    
    .exit:hover {
        background: var(--lightBlue);
        color: var(--mainDark)
    }
`;

export default ModalContainer;
