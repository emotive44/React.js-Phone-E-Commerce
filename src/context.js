import React from 'react';
import { storeProducts, detailProduct } from './data';

const { Consumer: ProductConsumer, Provider} = React.createContext();

class ProductProvider extends React.Component {
    state = {
        cart: [],
        products: [],
        modalProduct: [],
        detailProduct: [],
        modalOpen: false,
        cartSubTotal: 0,
        cartTotal: 0,
        cartTax: 0
    }

    componentDidMount() {
        this.setProducts();
    }

    setProducts = () => {
        let products = [];
        storeProducts.map(item => {
            const singleItem = {...item};
            products = [...products, singleItem];
        });
        this.setState({products})
    }

    getItem = id => {
        const product = this.state.products.find(item => item.id === id);
        return product;
    }

    handleDetail = (id) => {
        const product = this.getItem(id);
        this.setState({detailProduct: product})
    }

    addToCart = (id) => {
        let tempProducts = [...this.state.products];
        const index = tempProducts.indexOf(this.getItem(id));
        const product = tempProducts[index];
        product.inCart = true;
        product.count = 1;
        product.total = product.price;

        this.setState((prev) => ({
            products: tempProducts, 
            cart:[...prev.cart, product]
        }), () => { this.addTotals() });
    }

    openModal = id => {
        const product = this.getItem(id);
        this.setState({modalProduct: product, modalOpen: true});
    } 

    closeModal = () => {
        this.setState({modalOpen: false, modalProduct: []});
    }

    increment = id => {
        let tempCart = [...this.state.cart];
        let selectedProduct = this.getItem(id);

        selectedProduct.count += 1;
        selectedProduct.total = selectedProduct.count * selectedProduct.price;
        this.setState({cart: [...tempCart]}, () => this.addTotals())
    }

    decrement = id => {
        let tempCart = [...this.state.cart];
        let selectedProduct = this.getItem(id);
        
        selectedProduct.count -= 1;
        selectedProduct.total = selectedProduct.count * selectedProduct.price;
        this.setState({cart: [...tempCart]}, () => this.addTotals())
        
        if(selectedProduct.count === 0) {
            this.removeItem(id);
        }
    }

    removeItem = id => {
        let tempCart = [...this.state.cart];
        let tempProducts = [...this.state.products];

        tempCart = tempCart.filter(product => product.id !== id);

        let productsWithResetItem = tempProducts.map(item => { 
            if(item.id === id) {
                item.inCart = false;
                item.count = 0;
                item.total = 0;
             return item;
            }
            return item;
        })
        this.setState({cart: tempCart, products: productsWithResetItem}, () => this.addTotals());
    }

    clearCart = () => {
        let products = [...this.state.products];
        products.map(item => {
            item.inCart = false;
        })
        this.setState({cart: [], products}, () => this.addTotals());
    }

    addTotals = () => {
        let subTotal = 0;
        this.state.cart.map(item => {
            subTotal += item.total;
        });
        const tempTax = subTotal * 0.1;
        const tax  = parseFloat(tempTax.toFixed(2));
        const total = subTotal + tax;
        this.setState({cartSubTotal: subTotal, cartTax: tax, cartTotal: total});
    }

    render() {
        return (
            <Provider value={{
                ...this.state,
                addToCart: this.addToCart,
                openModal: this.openModal,
                closeModal: this.closeModal,
                handleDetail: this.handleDetail,
                removeItem: this.removeItem,
                increment: this.increment,
                decrement: this.decrement,
                clearCart: this.clearCart
            }}>
                {this.props.children}
            </Provider>
        )
    }
}

export { ProductProvider, ProductConsumer };
