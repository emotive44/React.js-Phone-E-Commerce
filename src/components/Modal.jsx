import React from 'react';
import styled from 'styled-components';
import { ProductConsumer } from '../context';
import ButtonContainer from './styles/Button';
import { Link } from 'react-router-dom';

class Modal extends React.Component {
    render() {
        return (
            <ProductConsumer>
                {
                    data => {
                        const { modalOpen, closeModal } = data;
                        const { img, title, price } = data.modalProduct;
                        if(!modalOpen) { return null };
                        return (
                            <ModalContainer>
                                <div className="container">
                                    <div className="row">
                                        <div id="modal" className="col-8 mx-auto col-md-6 col-lg-4 text-center text-capitalize ">
                                            <h5>item added to the cart</h5>
                                            <button className="exit" onClick={closeModal}>X</button>
                                            <img src={img} alt="phone" className="img-fluid"/>
                                            <h5>{title}</h5>
                                            <h5 className="text-muted">Price : ${price}</h5>
                                            <Link to='/'>
                                                <ButtonContainer onClick={closeModal}>
                                                    Continue Shopping
                                                </ButtonContainer>
                                            </Link>
                                            <Link to='/store'>
                                                <ButtonContainer cart onClick={closeModal}>
                                                    Go To Cart
                                                </ButtonContainer>
                                            </Link>
                                        </div>
                                    </div>
                                </div>
                            </ModalContainer>
                        )
                    }
                }
            </ProductConsumer>
        )
    }
}

const ModalContainer = styled.div`
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    display: flex;
    position: fixed;
    align-items: center;
    justify-content: center;
    background: rgba(0, 0, 0, 0.5);

    #modal {
        background: var(--mainWhite);
        border-radius: 1rem;
    }  
    .exit {
        position: absolute;
        top:0;
        right: 0;
        background: var(--mainYellow);
        border: none;
        outline: none;
        transition: all 0.6s linear;
        padding: 0 0.5rem;
        border-radius: 0 0.5rem 0 0 ;
        font-size: 2rem;
    }
    
    .exit:hover {
        background: var(--lightBlue);
        color: var(--mainDark)
    }
`;

export default Modal;
