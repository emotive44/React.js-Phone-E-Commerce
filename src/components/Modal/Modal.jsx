import React from 'react';
import { ProductConsumer } from '../../context';
import ModalContainer from './ModalContainer';
import ButtonContainer from '../styles/Button';
import { Link } from 'react-router-dom';

class Modal extends React.Component {
    render() {
        return (
            <ProductConsumer>
                {
                    data => {
                        const { modalOpen, closeModal } = data;
                        const { img, title, price } = data.modalProduct;
                        if(!modalOpen) { return null };
                        return (
                            <ModalContainer>
                                <div className="container">
                                    <div className="row">
                                        <div id="modal" className="col-8 mx-auto col-md-6 col-lg-4 text-center text-capitalize ">
                                            <h5>item added to the cart</h5>
                                            <button className="exit" onClick={closeModal}>X</button>
                                            <img src={img} alt="phone" className="img-fluid"/>
                                            <h5>{title}</h5>
                                            <h5 className="text-muted">Price : ${price}</h5>
                                            <Link to='/'>
                                                <ButtonContainer onClick={closeModal}>
                                                    Continue Shopping
                                                </ButtonContainer>
                                            </Link>
                                            <Link to='/store'>
                                                <ButtonContainer cart onClick={closeModal}>
                                                    Go To Cart
                                                </ButtonContainer>
                                            </Link>
                                        </div>
                                    </div>
                                </div>
                            </ModalContainer>
                        )
                    }
                }
            </ProductConsumer>
        )
    }
}

export default Modal;
