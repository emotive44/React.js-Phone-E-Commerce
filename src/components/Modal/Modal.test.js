import React from 'react';
import Modal from './Modal';
import { ProductContext } from '../../context';
import renderer from 'react-test-renderer';

jest.mock('react-router-dom', () => ({
    Link: 'Link'
}));

describe('Modal Component', () => {
    test('should render null if data is undefiend', () => {
        const component = renderer.create(<Modal />);
        expect(component.toJSON()).toMatchSnapshot();
    })

    test('should render null if modalOpen is false', () => {
        const component = renderer.create(
            <ProductContext.Provider value={{modalOpen:false, modalProduct: []}}>
                <Modal />
            </ProductContext.Provider>
        );
        expect(component.toJSON()).toMatchSnapshot();
    })

    
    test('should render null if modalProduct is undefined', () => {
        const component = renderer.create(
            <ProductContext.Provider value={{modalOpen:false}}>
                <Modal />
            </ProductContext.Provider>
        );
        expect(component.toJSON()).toMatchSnapshot();
    })
})