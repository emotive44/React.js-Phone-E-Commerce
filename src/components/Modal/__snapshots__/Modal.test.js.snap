// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Component should render null if data is undefiend 1`] = `null`;

exports[`Modal Component should render null if modalOpen is false 1`] = `null`;

exports[`Modal Component should render null if modalProduct is undefined 1`] = `null`;
